compiler/defer.go

```diff
@@ -127,12 +128,22 @@ func (c *Compiler) emitDefer(frame *Frame, instr *ssa.Defer) {
 		deferFrame = c.builder.CreateInsertValue(deferFrame, value, i, "")
 	}
 
-	// Put this struct in an alloca.
-	alloca := c.builder.CreateAlloca(deferFrameType, "defer.alloca")
-	c.builder.CreateStore(deferFrame, alloca)
-	if c.NeedsStackObjects() {
-		c.trackPointer(alloca)
+	// Put this struct in an allocation.
+	var alloca llvm.Value
+	if instr.Block().Dominates(instr.Parent().Blocks[0]) {
```

Comment:
I've been thinking more about this, and I think it's not a complete solution. I think it's safe, but take the following code for example:

```go
func foo() {
    if a {
        return
    }
    defer bar()
}
```

As I understand dominance theory, block at `defer bar()` does not dominate the entry block and thus the defer would be heap allocated (while there is no good reason for that).

I'm pretty sure there must be an easy way to calculate whether some block can be executed more than once, but can't really think of one. This problem must be solved a long time ago as it is important for compiler optimizations.